feat: Add support for query rewrite in vector_store.search

franciscojavierarceo opened the pull request:
# What does this PR do?

Actualize query rewrite in search API, add `default_query_expansion_model` and `query_expansion_prompt` in `VectorStoresConfig`.

Makes `rewrite_query` parameter functional in vector store search.
  - `rewrite_query=false` (default): Use original query
  - `rewrite_query=true`: Expand query via LLM, or fail gracefully if no LLM available

Adds 4 parameters to`VectorStoresConfig`:
- `default_query_expansion_model`: LLM model for query expansion (optional)
- `query_expansion_prompt`: Custom prompt template (optional, uses built-in default)
- `query_expansion_max_tokens`: Configurable token limit (default: 100)
- `query_expansion_temperature`: Configurable temperature (default: 0.3)

Enabled `run.yaml`:
```yaml
  vector_stores:
    rewrite_query_params:
      model:
        provider_id: "ollama"
        model_id: "llama3.2:3b-instruct-fp16"
      # prompt defaults to built-in
      # max_tokens defaults to 100
      # temperature defaults to 0.3
```

  Fully customized `run.yaml`:
```yaml
  vector_stores:
    default_provider_id: faiss
    default_embedding_model:
      provider_id: sentence-transformers
      model_id: nomic-ai/nomic-embed-text-v1.5
    rewrite_query_params:
      model:
        provider_id: ollama
        model_id: llama3.2:3b-instruct-fp16
      prompt: "Rewrite this search query to improve retrieval results by expanding it with relevant synonyms and related terms: {query}"
      max_tokens: 100
      temperature: 0.3
```

## Test Plan
Added test and recording

Example script as well:

```python
import asyncio
from llama_stack_client import LlamaStackClient
from io import BytesIO

def gen_file(client, text: str=""):
    file_buffer = BytesIO(text.encode('utf-8'))
    file_buffer.name = "my_file.txt"

    uploaded_file = client.files.create(
        file=file_buffer,
        purpose="assistants"
    )
    return uploaded_file

async def test_query_rewriting():
    client = LlamaStackClient(base_url="http://0.0.0.0:8321/")
    uploaded_file = gen_file(client, "banana banana apple")
    uploaded_file2 = gen_file(client, "orange orange kiwi")

    vs = client.vector_stores.create()
    xf_vs = client.vector_stores.files.create(vector_store_id=vs.id, file_id=uploaded_file.id)
    xf_vs1 = client.vector_stores.files.create(vector_store_id=vs.id, file_id=uploaded_file2.id)
    response1 = client.vector_stores.search(
                vector_store_id=vs.id,
                query="apple",
                max_num_results=3,
                rewrite_query=False
            )
    response2 = client.vector_stores.search(
                vector_store_id=vs.id,
                query="kiwi",
                max_num_results=3,
                rewrite_query=True,
            )

    print(f"\n🔵 Response 1 (rewrite_query=False):\n\033[94m{response1}\033[0m")
    print(f"\n🟢 Response 2 (rewrite_query=True):\n\033[92m{response2}\033[0m")

    for f in [uploaded_file.id, uploaded_file2.id]:
        client.files.delete(file_id=f)
    client.vector_stores.delete(vector_store_id=vs.id)

if __name__ == "__main__":
    asyncio.run(test_query_rewriting())
```

And see the screen shot of the server logs showing it worked. 
<img width="1111" height="826" alt="Screenshot 2025-11-19 at 1 16 03 PM" src="https://github.com/user-attachments/assets/2d188b44-1fef-4df5-b465-2d6728ca49ce" />

Notice the log:
```bash
 Query rewritten:
         'kiwi' → 'kiwi, a small brown or green fruit native to New Zealand, or a person having a fuzzy brown outer skin similar in appearance.'
```
So `kiwi` was expanded.

franciscojavierarceo commented on src/llama_stack/providers/utils/memory/vector_store.py:
```diff
@@ -333,3 +343,78 @@ async def query_chunks(
+        llm_models = [m for m in models_response.data if m.model_type == ModelType.llm]
+
+        # Filter out models that are known to be embedding models (misclassified as LLM)
+        embedding_model_patterns = ["minilm", "embed", "embedding", "nomic-embed"]
```
removing this and provider_priority below

mattf commented:
@franciscojavierarceo please update the description with the new proposed config and user interaction

mattf commented on src/llama_stack/core/datatypes.py:
```diff
@@ -376,6 +376,14 @@ class VectorStoresConfig(BaseModel):
+    default_query_expansion_model: QualifiedModel | None = Field(
+        default=None,
+        description="Default LLM model for query expansion/rewriting in vector search.",
+    )
```
```suggestion
    query_expansion_model: QualifiedModel | None = Field(
        default=None,
        description="LLM model for query expansion/rewriting in vector search.",
    )
```

there are other changes that need to go with this

mattf commented on src/llama_stack/core/resolver.py:
```diff
@@ -374,6 +374,13 @@ async def instantiate_provider(
 
+        # Add vector_stores_config for vector_io providers
+        if (
+            "vector_stores_config" in inspect.signature(getattr(module, method)).parameters
```
first impl of this should not let the user override the vector store configuration

mattf reviewed (commented):
#### complexity

for now, the rewrite model and rewrite prompt can be handled like the embedding model. you can configure them during stack startup and use them in the common impl w/o having to wire them into all the adapters.

later, it'll be good to get the special case handling of vector store config out of core/stack.py

#### spec

the spec doesn't mention letting api callers override the vector store config, but the implementation appears to let them. is this intended?

imho, the first impl shouldn't let users slot in their own models or prompts.

#### validation

the validation logic for performing the rewrite is split across two function, please unify.

the model validation logic should be left to the inference api. if the model is of the wrong type the inference api should handle the error generation.

vector store model type validation should happen during startup, if at all. right now the admin sees a successful startup and the user sees 400 errors about a misconfiguration.

cc @anik120

anik120 commented:
Ack @mattf - +1 with the startup validation approach.

@franciscojavierarceo FYI I created #4184 to tackle startup validation for optional dependencies so admins catch
misconfigurations at deploy time rather than users hitting 400 errors and needing to debug after. Looks like the validation changes here would align with that pattern.

On the production vs. R&D discussion: Matt's suggestion to use the "update vector store" path seems like a good middle
ground - gives R&D the flexibility to iterate without per-query instability. Happy to discuss additional validation patterns in
#4184 if it helps.

raghotham commented on src/llama_stack/core/datatypes.py:
```diff
@@ -376,6 +377,22 @@ class VectorStoresConfig(BaseModel):
         default=None,
         description="Default embedding model configuration for vector stores.",
     )
+    default_query_expansion_model: QualifiedModel | None = Field(
```
maybe just stick to one of expansion or rewrite across the board? its a little confusing if we have expansion here but rewrite elsewhere

raghotham commented on src/llama_stack/distributions/ci-tests/run-with-postgres-store.yaml:
```diff
@@ -288,5 +288,15 @@ vector_stores:
   default_embedding_model:
     provider_id: sentence-transformers
     model_id: nomic-ai/nomic-embed-text-v1.5
+  query_expansion_prompt: 'Expand this query with relevant synonyms and related terms.
```
no default query expansion model here?

franciscojavierarceo replied:
We don't default an inference model so I thought it's best to leave it as opt in and choosing a default LLM is actually not something we really do explicitly today (we do probably in a suboptimal way)

franciscojavierarceo commented on src/llama_stack/core/stack.py:
```diff
@@ -144,35 +144,62 @@ async def validate_vector_stores_config(vector_stores_config: VectorStoresConfig
     default_embedding_model = vector_stores_config.default_embedding_model
-    if default_embedding_model is None:
-        return
+    if default_embedding_model is not None:
```
done!

franciscojavierarceo commented:
>if rewrite is requested but the feature isn't configured, the user should get a 400 instead of silently ignoring

Will do.

>the globals are risky and will lead to bugs. there may already be one for two vector stores configured w/ different prompts. can we skip the globals and use something like self.vector_store.vector_stores_config in query_chunks instead?

@mattf if we want to use `self.vector_store.vector_stores_config` we'll have to modify `VectorStoreWithIndex` which will result in requiring us to modify all of the adapters. The globals are the alternative I thought made sense, let me know if you have alternative suggestions.

franciscojavierarceo commented:
@mattf I updated the code to handle the case where rewrite is requested but not configured.

Mind if I split the Adapter changes in a subsequent PR?

mattf commented:
> > if rewrite is requested but the feature isn't configured, the user should get a 400 instead of silently ignoring
> 
> Will do.
> 
> > the globals are risky and will lead to bugs. there may already be one for two vector stores configured w/ different prompts. can we skip the globals and use something like self.vector_store.vector_stores_config in query_chunks instead?
> 
> @mattf if we want to use `self.vector_store.vector_stores_config` we'll have to modify `VectorStoreWithIndex` which will result in requiring us to modify all of the adapters. The globals are the alternative I thought made sense, let me know if you have alternative suggestions.

my mistake, you're right that query_chunks is adapter specific. i was trying to suggest lifting the handling out of the individual adapters entirely. i don't see why adapters need to even know that the user is requesting query rewriting. the adapter just accepts a string to query against. worst case, the rewrite could happen in the mixin?

btw, i noticed that the routing layer only ever passes down a query: str, but the adapters accept a query: InterleavedContent.

mattf commented on src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:
```diff
@@ -85,14 +89,56 @@ def __init__(
+            temperature=self.vector_stores_config.rewrite_query_params.temperature or 0.3,
+        )
+
+        response = await self.inference_api.openai_chat_completion(request)  # type: ignore
```
openai_chat_completion can throw exceptions. they may include config details of the service that should not be exposed to users, e.g. the model being used, api credentials. a safer approach is to catch and log the detailed exception then send the user a 500 about query_rewrite failing.

franciscojavierarceo commented on src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:
```diff
@@ -85,14 +89,56 @@ def __init__(
+            temperature=self.vector_stores_config.rewrite_query_params.temperature or 0.3,
+        )
+
+        response = await self.inference_api.openai_chat_completion(request)  # type: ignore
```
to avoid adding the `inference_api` to the `init()` of all of the adapters. 

since you wanted to modifying the adapters I thought this was a reasonable compromise. LMK if you'd prefer I add them.

franciscojavierarceo replied:
it gets set during the provider instantiation using the `get_provider_impl()`, IMO the ideal thing to do is just add it to the init but I can clean that up in a follow up PR or a separate one if you'd like.

franciscojavierarceo commented on src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:
```diff
@@ -85,14 +89,56 @@ def __init__(
+        # Use custom prompt from config if provided, otherwise use built-in default
+        # Users only need to configure the model - prompt is automatic with optional override
+        custom_prompt = self.vector_stores_config.rewrite_query_params.prompt
+        if custom_prompt:
```
will remove
